ee/experimental/incident_utils.py

```diff
@@ -1,10 +1,152 @@
+
+    alerts = get_last_alerts(tenant_id, limit=use_n_historical_alerts, upper_timestamp=alert_upper_timestamp, lower_timestamp=alert_lower_timestamp)
+    incidents, _ = get_last_incidents(tenant_id, limit=use_n_hist_incidents, upper_timestamp=incident_upper_timestamp, lower_timestamp=incident_lower_timestamp)
+    nc_queue = NodeCandidateQueue()
+    
+    for candidate in [NodeCandidate(alert.fingerprint, alert.timestamp) for alert in alerts]:
+        nc_queue.push_candidate(candidate)
+    candidates = nc_queue.get_candidates()          
+    
+    graph = create_graph(tenant_id, [candidate.fingerprint for candidate in candidates], pmi_threshold, knee_threshold)
+    ids = []
+    
+    for component in nx.connected_components(graph):
+        if len(component) > min_incident_size:            
+            alerts_appended = False
+            for incident in incidents:
+                incident_fingerprints = set([alert.fingerprint for alert in incident.Incident.alerts])        
+                intersection = incident_fingerprints.intersection(component)
+        
+                if len(intersection) / len(component) >= incident_similarity_threshold:
+                    alerts_appended = True
+                    for alert in [alert for alert in alerts if alert.fingerprint in component]:
+                        if not is_alert_assigned_to_incident(alert.id, incident.Incident.id, tenant_id):
+                            assign_alert_to_incident(alert.id, incident.Incident.id, tenant_id)
+                    
+                    summary = generate_incident_summary(incident.Incident)
+                    update_incident_summary(incident.Incident.id, summary)
+                    
+            if not alerts_appended:
+                incident = create_incident_from_dict(tenant_id, {"name": "New Incident", "description": "Summorization is Disabled", "is_predicted": True})
+                ids.append(incident.id)
+                for alert in [alert for alert in alerts if alert.fingerprint in component]:
+                    if not is_alert_assigned_to_incident(alert.id, incident.id, tenant_id):
+                        assign_alert_to_incident(alert.id, incident.id, tenant_id)
+                    
+                summary = generate_incident_summary(incident)
+                update_incident_summary(incident.id, summary)
+    
+    pusher_client = get_pusher_client()
+    pusher_client.trigger(
+        f"private-{tenant_id}",
+        "ai-logs-change",
+        {"log": ALGORITHM_VERBOSE_NAME + " successfully executed."},
+    )
+    logger.info(
+        "Client notified on new AI log",
+        extra={"tenant_id": tenant_id},
+    )
+                
 
-from keep.api.models.db.alert import Alert
+    return {"incidents": [get_incident_by_id(tenant_id, incident_id) for incident_id in ids]}
 
 
 def mine_incidents(alerts: List[Alert], incident_sliding_window_size: int=6*24*60*60, statistic_sliding_window_size: int=60*60, 
```

Comment:
general note: worth moving all those consts from function def and explaining them